lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -271,6 +271,54 @@ static LogicalResult verifyComponentOp(ComponentOp op) {
                              "'calyx.wires', 'calyx.control'.";
 }
 
+void ComponentOp::build(OpBuilder &builder, OperationState &result,
+                        StringAttr name, ArrayRef<ComponentPortInfo> ports) {
+  using namespace mlir::function_like_impl;
+
+  result.addAttribute(::mlir::SymbolTable::getSymbolAttrName(), name);
+
+  SmallVector<Type, 4> inPortTypes, outPortTypes;
+  SmallVector<Attribute, 4> inPortNames, outPortNames;
+
+  for (auto &&port : ports) {
+    if (port.direction == PortDirection::INPUT) {
+      inPortTypes.push_back(port.type);
+      inPortNames.push_back(port.name);
+    } else {
+      outPortTypes.push_back(port.type);
+      outPortNames.push_back(port.name);
+    }
+  }
+
+  // Build the function type of the component.
+  auto functionType = builder.getFunctionType(inPortTypes, outPortTypes);
+  result.addAttribute(getTypeAttrName(), TypeAttr::get(functionType));
+
+  // Record the port names of the component.
+  result.addAttribute("inPortNames", builder.getArrayAttr(inPortNames));
+  result.addAttribute("outPortNames", builder.getArrayAttr(outPortNames));
+
+  // Create a single-blocked region.
+  result.addRegion();
+  Region *regionBody = result.regions[0].get();
+  Block *block = new Block();
+  regionBody->push_back(block);
+
+  // Add input ports to the body block.
+  for (auto port : ports) {
+    if (port.direction == PortDirection::OUTPUT)
+      continue;
+    block->addArgument(port.type);
+  }
+
+  // Insert the WiresOp and ControlOp.
+  auto ip = builder.saveInsertionPoint();
+  builder.setInsertionPointToStart(block);
+  builder.create<WiresOp>(result.location);
+  builder.create<ControlOp>(result.location);
+  builder.restoreInsertionPoint(ip);
```

Comment:
MLIR has something for this common pattern: https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/Builders.h#L282-L283.

Using it would look something like the following, and you wouldn't need to worry about resetting anything, since the guard will go out of scope at the end of the function, after those two `create` calls:

```
InsertionGuard guard(builder);
builder.setInsertionPoint...();
builder.create<>(...);
```

Reply:
This seems cleaner to me as well. Thanks for pointing it out.